share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java

```diff
@@ -17,58 +17,553 @@
+        leaderEpochMap.put(coordinatorKey, leaderEpoch);
+
+        // Returning the successfully retrieved snapshot value
+        return ReadShareGroupStateResponse.toResponseData(topicId, partition, offsetValue.startOffset(), offsetValue.stateEpoch(), stateBatches);
+    }
+
+    private Optional<CoordinatorResult<WriteShareGroupStateResponseData, CoordinatorRecord>> maybeGetWriteStateError(
+        WriteShareGroupStateRequestData request
+    ) {
+        String groupId = request.groupId();
+        WriteShareGroupStateRequestData.WriteStateData topicData = request.topics().get(0);
+        WriteShareGroupStateRequestData.PartitionData partitionData = topicData.partitions().get(0);
+
+        Uuid topicId = topicData.topicId();
+        int partitionId = partitionData.partition();
+
+        if (topicId == null) {
+            return Optional.of(getWriteErrorResponse(Errors.INVALID_TOPIC_EXCEPTION, null, null, partitionId));
+        }
+
+        if (partitionId < 0) {
+            return Optional.of(getWriteErrorResponse(Errors.INVALID_PARTITIONS, null, topicId, partitionId));
+        }
+
+        SharePartitionKey mapKey = SharePartitionKey.getInstance(groupId, topicId, partitionId);
+        if (leaderEpochMap.containsKey(mapKey) && leaderEpochMap.get(mapKey) > partitionData.leaderEpoch()) {
+            return Optional.of(getWriteErrorResponse(Errors.FENCED_LEADER_EPOCH, null, topicId, partitionId));
+        }
+        if (stateEpochMap.containsKey(mapKey) && stateEpochMap.get(mapKey) > partitionData.stateEpoch()) {
+            return Optional.of(getWriteErrorResponse(Errors.FENCED_STATE_EPOCH, null, topicId, partitionId));
+        }
+        if (metadataImage != null && (metadataImage.topics().getTopic(topicId) == null ||
+            metadataImage.topics().getPartition(topicId, partitionId) == null)) {
+            return Optional.of(getWriteErrorResponse(Errors.UNKNOWN_TOPIC_OR_PARTITION, null, topicId, partitionId));
+        }
+
+        return Optional.empty();
+    }
+
+    private Optional<ReadShareGroupStateResponseData> maybeGetReadStateError(ReadShareGroupStateRequestData request, Long offset) {
+        String groupId = request.groupId();
+        ReadShareGroupStateRequestData.ReadStateData topicData = request.topics().get(0);
+        ReadShareGroupStateRequestData.PartitionData partitionData = topicData.partitions().get(0);
+
+        Uuid topicId = topicData.topicId();
+        int partitionId = partitionData.partition();
+
+        if (topicId == null) {
+            return Optional.of(ReadShareGroupStateResponse.toErrorResponseData(
+                null, partitionId, Errors.INVALID_TOPIC_EXCEPTION, Errors.INVALID_TOPIC_EXCEPTION.message()));
+        }
+
+        if (partitionId < 0) {
+            return Optional.of(ReadShareGroupStateResponse.toErrorResponseData(
+                topicId, partitionId, Errors.INVALID_PARTITIONS, Errors.INVALID_PARTITIONS.message()));
```

Comment:
Those error messages are pretty generic. We should be more specific here.

Reply:
I'll go with invalid_request and custom message then